File: apps/api/app/schemas/mcp.py
"""MCP token / tool schemas — ADR-019."""

from __future__ import annotations

import uuid
from datetime import UTC, datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

McpScope = Literal["mcp:read"]
McpTokenStatus = Literal["active", "expired", "revoked"]


def default_mcp_scopes() -> list[McpScope]:
    return ["mcp:read"]


class McpTokenIssueRequest(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    expires_at: datetime | None = None
    scopes: list[McpScope] = Field(default_factory=default_mcp_scopes)

    @field_validator("name")
    @classmethod
    def _strip_name(cls, value: str) -> str:
        stripped = " ".join(value.split())
        if not stripped:
            raise ValueError("name은 공백만 입력할 수 없습니다.")
        return stripped

    @field_validator("expires_at")
    @classmethod
    def _future_expiry(cls, value: datetime | None) -> datetime | None:
        if value is None:
            return None
        normalized = value if value.tzinfo else value.replace(tzinfo=UTC)
        if normalized <= datetime.now(UTC):
            raise ValueError("expires_at은 현재 시각 이후여야 합니다.")
        return normalized


class AdminMcpTokenIssueRequest(McpTokenIssueRequest):
    user_id: uuid.UUID
    access_reason: str = Field(min_length=1, max_length=500)


class McpTokenRevokeRequest(BaseModel):
    access_reason: str = Field(min_length=1, max_length=500)


class McpTokenResponse(BaseModel):
    token_id: uuid.UUID
    user_id: uuid.UUID | None = None
    name: str
    scopes: list[McpScope]
    masked_token: str
    expires_at: datetime | None
    last_used_at: datetime | None
    revoked_at: datetime | None
    created_at: datetime


class McpTokenIssueResponse(McpTokenResponse):
    token: str


class McpToolDescriptor(BaseModel):
    name: str
    description: str
    inputSchema: dict[str, Any]


class McpToolCallRequest(BaseModel):
    arguments: dict[str, Any] = Field(default_factory=dict)


class McpToolCallResponse(BaseModel):
    tool: str
    result: dict[str, Any]

File: packages/schemas/src/mcp.ts
import { z } from 'zod';
import { Iso8601Schema } from './common';

export const McpScopeSchema = z.literal('mcp:read');
export type McpScope = z.infer<typeof McpScopeSchema>;

export const McpTokenStatusSchema = z.enum(['active', 'expired', 'revoked']);
export type McpTokenStatus = z.infer<typeof McpTokenStatusSchema>;

export const McpTokenIssueRequestSchema = z.object({
  name: z.string().min(1).max(120),
  expires_at: Iso8601Schema.nullable().optional(),
  scopes: z.array(McpScopeSchema).optional().default(['mcp:read']),
});
export type McpTokenIssueRequest = z.infer<typeof McpTokenIssueRequestSchema>;

export const AdminMcpTokenIssueRequestSchema = McpTokenIssueRequestSchema.extend({
  user_id: z.string().uuid(),
  access_reason: z.string().min(1).max(500),
});
export type AdminMcpTokenIssueRequest = z.infer<typeof AdminMcpTokenIssueRequestSchema>;

export const McpTokenRevokeRequestSchema = z.object({
  access_reason: z.string().min(1).max(500),
});
export type McpTokenRevokeRequest = z.infer<typeof McpTokenRevokeRequestSchema>;

export const McpTokenSchema = z.object({
  token_id: z.string().uuid(),
  user_id: z.string().uuid().nullable().optional(),
  name: z.string(),
  scopes: z.array(McpScopeSchema),
  masked_token: z.string(),
  expires_at: Iso8601Schema.nullable(),
  last_used_at: Iso8601Schema.nullable(),
  revoked_at: Iso8601Schema.nullable(),
  created_at: Iso8601Schema,
});
export type McpToken = z.infer<typeof McpTokenSchema>;

export const McpTokenIssueResponseSchema = McpTokenSchema.extend({
  token: z.string().min(1),
});
export type McpTokenIssueResponse = z.infer<typeof McpTokenIssueResponseSchema>;
